implementations/rust/examples/credentials/src/issuer.rs

```diff
@@ -2,32 +2,57 @@ mod util;
 
 use ockam::{CredentialAttribute, CredentialIssuer};
 use std::collections::{BTreeMap, BTreeSet};
+use std::convert::TryFrom;
 use std::io::Write;
-use std::os::unix::net::{UnixListener, UnixStream};
-use util::{example_schema, CredentialMessage};
+use std::path::PathBuf;
+use structopt::StructOpt;
+use util::{example_schema, CredentialMessage, Listener, Stream};
+
+#[derive(StructOpt)]
+struct Args {
+    /// HEX encoded issuer signing key
+    #[structopt(long)]
+    secret_key: Option<String>,
+    #[structopt(long)]
+    port: Option<usize>,
+    #[structopt(long, parse(from_os_str))]
+    socket: Option<PathBuf>,
+}
 
 fn main() {
+    let args = Args::from_args();
+
     // Create a new issuer.
     // CredentialIssuer has a credential signature public key.
     // CredentialIssuer creates a proof of possession
     // so users can verify it.
     // These should be posted such that verifiers and
     // holders can check them
-    let issuer = CredentialIssuer::new();
+    let issuer = if args.secret_key.is_none() {
+        CredentialIssuer::new()
+    } else {
+        let sk = <[u8; 32]>::try_from(
+            hex::decode(args.secret_key.as_ref().unwrap())
+                .unwrap()
+                .as_slice(),
+        )
+        .unwrap();
+        CredentialIssuer::with_signing_key(sk)
+    };
     let pk = issuer.get_public_key();
     let pop = issuer.create_proof_of_possession();
     let schema = example_schema();
     let mut pending_offers = BTreeSet::new();
 
     //TODO: publish public key and proof of possession for verifiers
-
-    let listener = UnixListener::bind("/tmp/issuer.socket").unwrap();
+    let listener = Listener::bind(args.socket, args.port).unwrap();
```

Comment:
It better to use .expect("message") than .unwrap(), because it much easier to reason about error with explanation.

Reply:
The Result Err is a String. It prints either way.

Reply:
unwrap() will print error message, but it will not explain reason. Usually, in this case I use something like 

.expect("Cannot bind socket to address. Maybe another daemon is already running at this ip:port? Try to close daemon, or use another ip:port combination, e.g. 127.0.0.2:9090.");

It serves both as the comment in code and as debugging message for a system administrator.

Reply:
The example code underneath is already doing that. Hence the unwrap